controller/linodeobjectstoragebucket_controller_test.go

```diff
@@ -0,0 +1,184 @@
+		// Create a new gomock controller for each test run
+		mockCtrl = gomock.NewController(GinkgoT())
+	})
+
+	AfterEach(func() {
+		// At the end of each test run, tell the gomock controller it's done
+		// so it can check configured expectations and validate the methods called
+		mockCtrl.Finish()
+	})
+
+	It("should reconcile an object apply", func() {
+		mockClient := mock.NewMockLinodeObjectStorageClient(mockCtrl)
+
+		getCall := mockClient.EXPECT().
+			GetObjectStorageBucket(gomock.Any(), obj.Spec.Cluster, gomock.Any()).
+			Return(nil, nil).
+			Times(1)
+
+		createBucketCall := mockClient.EXPECT().
+			CreateObjectStorageBucket(gomock.Any(), gomock.Any()).
+			Return(&linodego.ObjectStorageBucket{
+				Label:    *obj.Spec.Label,
+				Cluster:  obj.Spec.Cluster,
+				Created:  util.Pointer(time.Now()),
+				Hostname: "hostname",
+			}, nil).
+			Times(1).
+			After(getCall)
+
+		for idx, permission := range []string{"rw", "ro"} {
+			mockClient.EXPECT().
+				CreateObjectStorageKey(
+					gomock.Any(),
+					gomock.Cond(func(opt any) bool {
+						createOpt, ok := opt.(linodego.ObjectStorageKeyCreateOptions)
+						if !ok {
+							return false
+						}
+
+						return createOpt.Label == fmt.Sprintf("%s-%s", *obj.Spec.Label, permission)
+					}),
+				).
+				Return(&linodego.ObjectStorageKey{ID: idx}, nil).
+				Times(1).
+				After(createBucketCall)
+		}
+
+		reconciler := &LinodeObjectStorageBucketReconciler{
+			Client:              k8sClient,
+			Scheme:              k8sClient.Scheme(),
+			Logger:              ctrl.Log.WithName("LinodeObjectStorageBucketReconciler"),
+			Recorder:            recorder,
+			LinodeClientBuilder: mockClientBuilder(mockClient),
+		}
+
+		objectKey := client.ObjectKeyFromObject(obj)
+		Expect(k8sClient.Create(ctx, obj)).To(Succeed())
+		_, err := reconciler.Reconcile(ctx, reconcile.Request{
+			NamespacedName: objectKey,
+		})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("updating its status fields")
+		Expect(k8sClient.Get(ctx, objectKey, obj)).To(Succeed())
+		Expect(*obj.Status.Hostname).To(Equal("hostname"))
+		Expect(*obj.Status.KeySecretName).To(Equal(secretName))
+		Expect(*obj.Status.LastKeyGeneration).To(Equal(*obj.Spec.KeyGeneration))
+		Expect(*obj.Status.LastKeyGeneration).To(Equal(0))
+		Expect(obj.Status.Ready).To(BeTrue())
+
+		By("creating a Secret with access keys")
+		Expect(k8sClient.Get(ctx, client.ObjectKey{
+			Name:      secretName,
+			Namespace: obj.Namespace,
+		}, &secret)).To(Succeed())
+		Expect(secret.Data["read_write"]).To(Not(BeNil()))
+		Expect(secret.Data["read_only"]).To(Not(BeNil()))
+
+		By("recording the expected event")
+		Expect(<-recorder.Events).To(ContainSubstring("Object storage bucket applied"))
+	})
+
+	It("should reconcile an object delete", func() {
+		mockClient := mock.NewMockLinodeObjectStorageClient(mockCtrl)
+
+		for i := range 2 {
+			mockClient.EXPECT().
+				DeleteObjectStorageKey(gomock.Any(), i).
+				Return(nil).
+				Times(1)
+		}
```

Comment:
Does gomock let us do this?
```suggestion
	mockClient.EXPECT().
		DeleteObjectStorageKey(gomock.Any(), i).
		Return(nil).
		Times(2)
```

Reply:
The mock recorder returned from `EXPECT()` counts each distinct function call (and not how many times a function is called), which is why we assert that it will call both `DeleteObjectStorageKey(any, 0)` and `DeleteObjectStorageKey(any, 1)` exactly once each.